py/kubeflow/testing/ci/kf_unittests.py

```diff
@@ -1,23 +1,248 @@
+    # The directory within the kubeflow_testing submodule containing
+    # py scripts to use.
+    self.kubeflow_testing_py = self.src_root_dir + "/kubeflow/testing/py"
+
+    self.go_path = self.test_dir
+
+  def _build_workflow(self):
+    """Create the scaffolding for the Argo workflow"""
+    workflow = {
+      "apiVersion": "argoproj.io/v1alpha1",
+      "kind": "Workflow",
+      "metadata": {
+        "name": self.name,
+        "namespace": self.namespace,
+        "labels": argo_build_util.add_dicts([{
+            "workflow": self.name,
+            "workflow_template": TEMPLATE_LABEL,
+          }, argo_build_util.get_prow_labels()]),
+      },
+      "spec": {
+        "entrypoint": E2E_DAG_NAME,
+        # Have argo garbage collect old workflows otherwise we overload the API
+        # server.
+        "ttlSecondsAfterFinished": 7 * 24 * 60 * 60,
+        "volumes": [
+          {
+            "name": "gcp-credentials",
+            "secret": {
+              "secretName": "kubeflow-testing-credentials",
+            },
+          },
+          {
+            "name": DATA_VOLUME,
+            "persistentVolumeClaim": {
+              "claimName": NFS_VOLUME_CLAIM,
+            },
+          },
+        ],
+        "onExit": EXIT_DAG_NAME,
+        "templates": [
+          {
+           "dag": {
+                "tasks": [],
+                },
+           "name": E2E_DAG_NAME,
+          },
+          {
+            "dag": {
+              "tasks": [],
+              },
+              "name": EXIT_DAG_NAME,
+            }
+        ],
+      },  # spec
+    } # workflow
+
+    return workflow
+
+  def _build_task_template(self):
+    """Return a template for all the tasks"""
+
+    task_template = {'activeDeadlineSeconds': 3000,
+     'container': {'command': [],
+      'env': [
+        {"name": "GOOGLE_APPLICATION_CREDENTIALS",
+         "value": "/secret/gcp-credentials/key.json"}
+       ],
+      'image': 'gcr.io/kubeflow-ci/test-worker:latest',
+      'imagePullPolicy': 'Always',
+      'name': '',
+      'resources': {'limits': {'cpu': '4', 'memory': '4Gi'},
+       'requests': {'cpu': '1', 'memory': '1536Mi'}},
+      'volumeMounts': [{'mountPath': '/mnt/test-data-volume',
+        'name': 'kubeflow-test-volume'},
+       {'mountPath': '/secret/gcp-credentials', 'name': 'gcp-credentials'}]},
+     'metadata': {'labels': {
+       'workflow_template': TEMPLATE_LABEL}},
+     'outputs': {}}
+
+    # Define common environment variables to be added to all steps
+    common_env = [
+      {'name': 'PYTHONPATH',
+       'value': ":".join([self.kubeflow_py, self.kubeflow_testing_py])},
+      {'name': 'GOPATH',
+        'value': self.go_path},
+      {'name': 'KUBECONFIG',
+       'value': os.path.join(self.test_dir, 'kfctl_test/.kube/kubeconfig')},
+    ]
+
+    task_template["container"]["env"].extend(common_env)
+
+    task_template = argo_build_util.add_prow_env(task_template)
+
+    return task_template
+
+  def build(self):
```

Comment:
If I have to create an argo workflow for another repository, seems I have to copy this unittest.py and modify only these specific tasks.

Is it better to move all these functions to argo_build_util.py; and have a smaller interface for user to use, i.e. only `newTask(name, cmd, deps)` and `newWorkflow(tasks, entry)`?

Reply:
@zhenghuiwang We can iterate over time. If you look at kubeflow/kubeflow#4148 which defines the E2E test for Argo workflow. I think there will be more variability between different E2E tests. So I'm not sure what the right abstraction/syntactic sugar is.

I think after we write a few more pipelines we will get more insight and can refine it.